lib/OpenQA/WebAPI/Controller/Step.pm

```diff
@@ -571,11 +594,11 @@ sub viewimg {
 
     my @needles;
     if ($module_detail->{needle}) {
-        my $needle = needle_info($module_detail->{needle}, $distribution, $dversion);
+        my $needle = needle_info($module_detail->{needle}, $distribution, $dversion, $module_detail->{file});
         if ($needle) {    # possibly missing/broken file
             my $info = {
                 name    => $module_detail->{needle},
-                image   => $self->needle_url($distribution, $module_detail->{needle} . '.png', $dversion),
+                image   => $self->needle_url($distribution, $module_detail->{needle} . '.png', $dversion, $needle->{json}),
                 areas   => $needle->{area},
                 matches => []};
             calc_matches($info, $module_detail->{area});
```

Comment:
This is really confusing IMO and I get you are not the original author, but can you change variable name `$needle` to `$needleinfo`.
In later chunk there is `$needle` var too, but this time I guess the one provided by module details hence one time `$needle->{json}` and another time `$needle->{file}` point to the same file.